[serve][LLM] Enable cross-node placement groups for serve.llm

nrghosh opened the pull request:
## Why are these changes needed?

- default strict-pack --> pack to unlock cross node tp/pp
- deprecate / remove `resources_per_bundle`


## Enable Multi-Node Tensor/Pipeline Parallelism for Ray Serve LLM

This PR enables Ray Serve LLM deployments to span multiple nodes when using tensor parallelism (TP) and/or pipeline parallelism (PP).

### Core Changes

**New `placement_group_config` API**
- Replaces `resources_per_bundle` with structured placement group configuration
- Supports custom resource bundles and placement strategies (PACK, SPREAD, STRICT_PACK, STRICT_SPREAD)
- Example:
  ```python
  placement_group_config = {
      "bundles": [{"GPU": 1, "CPU": 2}] * 8,
      "strategy": "PACK"
  }
  ```

**Default Placement Strategy: PACK**
- Changed from STRICT_PACK to PACK for all configurations (TP, PP, TP+PP)
- PACK enables cross-node placement when single-node resources are insufficient
- Data parallel deployments automatically override to STRICT_PACK (each replica must be co-located)

**Placement Group Configuration Flow**
- Custom bundles pass through validation (`PlacementGroupConfig`, `BundleConfig`)
- Accelerator type labels automatically injected into bundles
- First bundle merges with replica actor resources (CPU, GPU, memory)
- Resource validation occurs at placement group creation, not in Ray Serve

### Bug Fixes

**S3 Mirror Model Resolution**
- Fixes release test errors when loading models from S3 mirrors
- Root cause: vLLM queries HuggingFace API even after downloading from S3
- Solution: Conditionally resolve to local cache path only when `mirror_config` is present
- Preserves normal HF behavior for trust_remote_code validation

### Testing

**Unit Tests** (`test_multi_node_placement_groups.py`)
- Custom PG configurations with various TP/PP/strategy combinations
- Default PACK strategy validation
- Accelerator type injection and resource merging
- Data parallel STRICT_PACK override
- Bundle count validation for different parallelism configurations

**Integration Tests** (`test_llm_serve_multi_node_integration.py`)
- Multi-node TP+PP deployment (8 GPUs across 2 nodes)
- Data parallelism with placement group override
- Validates actual vLLM engine initialization
- Nightly release test `llm_serve_multi_node` on 2× g6.12xlarge (4 L4 GPUs each)
- Tests TP×PP=8 configurations that force cross-node placement

### Documentation

- Updated quick-start guide to use `placement_group_config` instead of `resources_per_bundle`
- Updated all test fixtures and examples to new API


### Related Work

Similar multi-node TP/PP support was added to `ray.data.llm` in [#56779](https://github.com/ray-project/ray/pull/56779), which introduced `placement_group_config` for vLLM processors in Ray Data batch inference workflows. Both efforts align on the same placement group configuration API and PACK-by-default strategy, enabling consistent multi-node LLM deployment patterns across Ray Serve (online inference) and Ray Data (batch inference).

## Related issue number

https://github.com/ray-project/ray/issues/55405

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

nrghosh committed:
Serve LLM: enable multi-node TP/PP with placement groups

- Default placement strategy to PACK (allow cross-node)
- Add PlacementGroupSchema/BundleSchema (float CPU/GPU types)
- Support custom placement_group_config; no Serve-side GPU<=1 enforcement
- Deprecate resources_per_bundle (kept for BC with warning)
- Keep accelerator label injection into bundles

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Serve LLM tests: PACK default and multi-node correctness

- Update unit tests for PACK-by-default behavior
- Include integration test for custom placement groups and cross-node TP/PP

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Serve LLM release: add multi-node TP/PP correctness run

- Add nightly job llm_serve_multi_node_tp_pp_correctness (reuse correctness suite)
- Lint

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
wip - serve.llm cross node tp/pp

- default PG strategy PACK
- add placement_group_config schema
- merge replica CPU/GPU into first bundle
- align engine PG creation with PACK default
- testing: validation, integration (with cleanup), release tests for multi-node tp/pp correctness

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
wip - fix example

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
wip - fix release tests to group-llm

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
wip

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
proper tests for new cross-node PGs

- llm_serve_multi_node_placement_groups to vlaidate  placement_group_config and PACK default
- llm_serve_multi_node_integration: test acutal multi-node deployments with custom PGs
- Use llm-serve group on `llm_four_L4_gpu_head_node.yaml` cluster

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
wip - test cleanup

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
wip

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
wip - cleanup

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
clean up example

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
address CR feedback

- handle redundant setdefault
- deduplicate placement_group_config logic
- more cleanup

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh commented on python/ray/llm/_internal/serve/configs/server_models.py:
```diff
@@ -178,6 +178,16 @@ class LLMConfig(BaseModelExtended):
+        default=None,
+        description=(
+            "Ray placement group configuration for scheduling vLLM engine workers. "
+            "Defines resource bundles and placement strategy for multi-node deployments. "
+            "Should contain 'bundles' (list of resource dicts) and optionally 'strategy' "
```
Here's why I think it's better to have a dict
1. Cleaner interface since we're passing PG configuration through multiple layers (LLMConfig --> VLLMEngineConfig --> engine creation)
2. Extensibility - if we need to add more options in the future, can just extend the dict rather than adding a new top-level field to LLMConfig
3. Easier to validate as a unit using `PlacementGroupSchema`

Overall I think it's just cleaner but if you feel strongly it's easy to refactor

nrghosh committed:
Merge branch 'master' into multi-node-serve-llm

nrghosh reviewed (commented):
looking into release test failure for [`llm_serve_llama_3dot2_1B_s3 (None) (0)`](https://buildkite.com/ray-project/release/builds/60753/steps/table?sid=01999c11-1cd0-458a-97f8-5d5e85c747ba)

> Error retrieving file list: 401 Client Error. (Request ID: Root=1-68dc3a7b-035e63806864b61966288ed0;4399140d-4d2f-4e67-8ed2-36be6069ec19)
ERROR 09-30 13:15:55 [config.py:278] 
ERROR 09-30 13:15:55 [config.py:278] Repository Not Found for url: https://huggingface.co/api/models/my_llama/tree/main?recursive=True&expand=False.
ERROR 09-30 13:15:55 [config.py:278] Please make sure you specified the correct `repo_id` and `repo_type`.
ERROR 09-30 13:15:55 [config.py:278] If you are trying to access a private or gated repo, make sure you are authenticated. For more details, see https://huggingface.co/docs/huggingface_hub/authentication
ERROR 09-30 13:15:55 [config.py:278] Invalid username or password., retrying 1 of 2
ERROR 09-30 13:15:57 [config.py:276] Error retrieving file list: 401 Client Error. (Request ID: Root=1-68dc3a7d-0217ebd62c1e648607c21af7;38e0575c-960f-43d3-ac3f-55874826e7d2)
ERROR 09-30 13:15:57 [config.py:276] 

**Note**: error was caused because vLLM was hitting huggingface for model metadata even after downloading from S3 mirror to local cache. Change was strict_pack->pack which was a different code-path and exposed an existing bug. Fixed by https://github.com/ray-project/ray/pull/56980/commits/5c44be8a1221cf6db97c22934e374fc0c8c1b0bd

ruisearch42 commented on python/ray/llm/_internal/serve/deployments/llm/vllm/vllm_models.py:
```diff
@@ -32,6 +32,28 @@
+        extra = "allow"  # Allow arbitrary resource types
+
+
+class PlacementGroupSchema(BaseModelExtended):
```
better to name it `PlacementGroupConfig` to be aligned with other fields?

ruisearch42 commented on python/ray/llm/_internal/serve/deployments/llm/vllm/vllm_models.py:
```diff
@@ -32,6 +32,28 @@
 logger = get_logger(__name__)
 
 
+class BundleSchema(BaseModelExtended):
```
`BundleConfig`?

kouroshHakha commented on python/ray/llm/_internal/serve/configs/server_models.py:
```diff
@@ -480,21 +492,59 @@ def _set_deployment_placement_options(self) -> Dict[str, Any]:
+                )[0]
+                pg_bundles[0] = merged_first_bundle
+
+            deployment_config.update(
+                {
+                    "placement_group_bundles": pg_bundles,
+                    "placement_group_strategy": strategy,
+                }
```
Yeah, something related to this: I think having deployment_config be part of LLMConfig was not a wise decision. First of all it is not explicit about which deployment, is it llm_server? Is it Router deployment? Etc

Second it creates this confusion about source of truth. This is one of the friction points I want to fix in a follow up PR and remove deployment_config setting entirely from LLMConfig.

ruisearch42 commented on python/ray/llm/_internal/serve/deployments/llm/vllm/vllm_models.py:
```diff
@@ -179,18 +222,40 @@ def num_devices(self) -> int:
+        if self.placement_group_config:
+            return self.placement_group_config.strategy
+        # Default to PACK  (cross-node best-effort placement)
+        # DP deployments overriden to STRICT_PACK in Serve config
```
nit: override or overridden

ruisearch42 commented on python/ray/llm/_internal/serve/deployments/llm/vllm/vllm_models.py:
```diff
@@ -32,6 +32,28 @@
+    Note: Counts are floats to align with Ray resource typing.
+    """
+
+    CPU: float = Field(default=1.0, ge=0.0, description="Number of CPUs per bundle")
```
default should be 0, right? previously we don't add CPU resources for vLLM worker

ruisearch42 commented on python/ray/llm/_internal/serve/deployments/llm/vllm/vllm_models.py:
```diff
@@ -248,11 +313,28 @@ def get_or_create_pg(self) -> PlacementGroup:
+                for bundle_schema in self.placement_group_config.bundles:
+                    bundle = bundle_schema.dict()
+                    if self.accelerator_type:
+                        bundle.setdefault(self.ray_accelerator_type(), 0.001)
```
this is done also in the method `placement_bundles()`. Can we call that method instead to reduce code duplication?

Actually you can just remove the if block from L317 to L324?

ruisearch42 reviewed (commented):
Overall LGTM

ruisearch42 reviewed (commented):
not a big concern, but wondering if we could merge some of the tests to be more concise, since the setup is pretty similar

ruisearch42 commented on release/llm_tests/serve/test_llm_serve_multi_node_integration.py:
```diff
@@ -0,0 +1,261 @@
+    # Cleanup handled by autouse fixture
+
+
+@pytest.mark.skip(
+    reason="Requires autoscaler/multi-node resources; covered in release jobs."
+)
```
I see you already used `llm_four_L4_gpu_head_node.yaml`, should we remove this?

ruisearch42 replied:
OK that's just 4xL4, so not enough for this.
Could you explain which release jobs cover this?

kouroshHakha replied:
+1

ruisearch42 reviewed (commented):
Otherwise LGTM. Can we add a test for DP to make sure there is no regression?

kouroshHakha commented on release/llm_tests/serve/test_llm_serve_multi_node_integration.py:
```diff
@@ -0,0 +1,261 @@
+        "strategy": "STRICT_PACK",
+    }
+
+    llm_config = LLMConfig(
```
+1

kouroshHakha reviewed (commented):
A couple of comments:

nrghosh committed:
Merge branch 'master' into multi-node-serve-llm

nrghosh committed:
address rui comments

- refactor BundleSchema/PlacementGroupSchema to BundleConfig/PlacementGroupConfig
- fix typo
- deduplicate if block in get_or_create_pg; use placement_bundles property
- deduplicate deployment_config.update logic in server_models.py

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
address rui/kourosh feedback

- Keep placement_group_config as Dict[str, Any] in both LLMConfig and VLLMEngineConfig
- Add field_validator in VLLMEngineConfig that validates via PlacementGroupConfig, then model_dump() back to dict
- Update placement_strategy and placement_bundles properties to work with dict (not model object)
- Simplify from_llm_config: no need to construct PlacementGroupConfig (validator handles it)
- Remove resources_per_bundle field entirely
- Change CPU default to 0.0
- Deduplicate PG creation and deployment config logic
- Rename Schema→Config classes
- Consolidate similar tests
- Add multi-GPU passthrough test

Release test improvements:
- Add llm_2x_2xl4.yaml
- Update llm_serve_multi_node_integration to use multi-node cluster
- Add comments explaining cross-node TP/PP test coverage

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
wip - multinode release test (simpler)

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
wip - lint

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh commented on python/ray/llm/_internal/serve/configs/server_models.py:
```diff
@@ -166,16 +166,19 @@ class LLMConfig(BaseModelExtended):
-        description="This will override the default resource bundles for placement groups. "
-        "You can specify a custom device label e.g. {'NPU': 1}. "
-        "The default resource bundle for LLM Stage is always a GPU resource i.e. {'GPU': 1}.",
+        description=f"The type of accelerator runs the model on. Only the following values are supported: {str([t.value for t in GPUType])}",
```
this is a minor concern about when validation happens - but the current code does eventually validate via VLLMEngineConfig, so it's not critical - the existing tests in `test_multi_node_placement_groups.py` show that the functionality works correctly

nrghosh committed:
wip - fix

- Empty placement_bundles handling:
- GPU detection fix
- Add DP deployment test, don't call get_serve_options() twice
- disable hf download for s3 release test (failing in CI)

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
bugfix - resolve S3 mirror models to local path (conditional on mirror_config)

Root Cause:
S3 mirror test fails with 404/401 HF API errors. vLLM queries huggingface.co
for model metadata even after downloading from S3 to local cache.

Why Our PR Exposed This:
Changed default placement strategy from STRICT_PACK to PACK for single-GPU configs.
- STRICT_PACK path: Downloads locally and updates hf_model_id to local path
- PACK path: Downloads on worker nodes but never updates hf_model_id
S3 test now takes PACK path, hits the bug.

Original Fix (was too broad):
Unconditionally resolved all models to local cache if present.
Broke TestRemoteCode tests which need to fetch from HF to validate trust_remote_code.

Refined Fix:
In _get_vllm_engine_config (GPU task before vLLM init):
- Only resolve if engine_config.mirror_config exists (intentional S3/GCS download)
- Call get_model_location_on_disk to get local snapshot path
- Update engine_config.hf_model_id so vLLM uses filesystem path
- If no mirror_config, preserve original model_id (normal HF behavior)

This fixes S3 mirror tests without breaking remote code validation tests.

Fixes
- llm_serve_llama_3dot2_1B_s3 (404/401 errors)
- TestRemoteCode::test_remote_code_success

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh commented on python/ray/llm/_internal/serve/deployments/llm/vllm/vllm_models.py:
```diff
@@ -179,18 +224,29 @@ def num_devices(self) -> int:
+                    bundle.setdefault(self.ray_accelerator_type(), 0.001)
+                bundles.append(bundle)
+            return bundles
+
```
Not a bug -- 

the  if-else ensures exactly one of these happens:
1. Empty bundles: pg_bundles = [replica_actor_resources] (direct assignment)
2. Non-empty bundles: merge replica resources into first bundle (single merge)

nrghosh reviewed (commented):
fixed bad merge + preexisting mirror download bug

nrghosh commented on python/ray/llm/_internal/serve/deployments/llm/vllm/vllm_engine.py:
```diff
@@ -55,6 +55,19 @@ def _get_vllm_engine_config(
+        local_path = get_model_location_on_disk(engine_config.actual_hf_model_id)
+        if local_path and local_path != engine_config.actual_hf_model_id:
+            engine_config.hf_model_id = local_path
+            logger.info(f"Resolved model from mirror to local path: {local_path}")
```
This is intentional -   `engine_config` cache preserves `hf_model_id` across calls so vLLM doesn't re-download. mutation is safe because it only happens once (when model is first resolved to local path) and subsequent calls reuse the cached value

nrghosh committed:
Fix test failures from merge

- Update test_llm_server.py to expect PACK default (not STRICT_PACK)
- Fix release test import to use public API (ray.serve.llm)
- Update test to use placement_group_config instead of deprecated resources_per_bundle

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

kouroshHakha commented on python/ray/llm/tests/serve/cpu/configs/test_multi_node_placement_groups.py:
```diff
@@ -0,0 +1,239 @@
+
+    # First bundle: merged replica actor resources
+    first_bundle = serve_options["placement_group_bundles"][0]
+    assert first_bundle["CPU"] >= 2
+    assert first_bundle["GPU"] >= 1
```
should we just assert for equality here?

kouroshHakha commented on python/ray/llm/tests/serve/cpu/configs/test_multi_node_placement_groups.py:
```diff
@@ -0,0 +1,239 @@
+    # Serve should accept and pass through GPU=2 to placement group
+    serve_options = llm_config.get_serve_options()
+    assert serve_options["placement_group_bundles"][0]["GPU"] == 2
+    assert serve_options["placement_group_bundles"][0]["CPU"] >= 4
```
we should assert for equality here? no?

kouroshHakha reviewed (commented):
some comments:

nrghosh commented on python/ray/llm/_internal/serve/configs/server_models.py:
```diff
@@ -24,8 +26,10 @@
 )
 from ray.llm._internal.common.utils.import_utils import try_import
 from ray.llm._internal.serve.configs.constants import (
+    DEFAULT_MAX_ONGOING_REQUESTS,
```
I see it still here? https://github.com/ray-project/ray/commit/8bfb12450c41de693c75a14f40665e4bac04e801#diff-f10cbf826427ee5705e130e69f8485ded67d5921c7dbb6452d8f452879385983R11-R25

nrghosh commented on python/ray/llm/_internal/serve/configs/server_models.py:
```diff
@@ -430,6 +437,180 @@ def update_engine_kwargs(self, **kwargs: Any) -> None:
         if self._engine_config:
             self._engine_config.engine_kwargs.update(kwargs)
 
+    def _merge_replica_actor_and_child_actor_bundles(
```
<img width="1139" height="334" alt="image" src="https://github.com/user-attachments/assets/d8731d08-d99b-4994-afd0-cbf005a1e111" />

still see it here? https://github.com/ray-project/ray/commit/8bfb12450c41de693c75a14f40665e4bac04e801#diff-89512bc3cc2837b9851be5c3e6c5082e92655ed85a41eace1a0d56a2c57c0896R61

nrghosh committed:
Merge remote-tracking branch 'upstream/master' into multi-node-serve-llm

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
docfix - use placement_group_config instead of resources_per_bundle

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
wip - re-do merge, fix issues and address final comments

- update tests to use get_deployment_options
- DP strict_pack override logic (in llm_server)
- test configs can use cpu:1
- tighter asserts in tests (instead of >=)

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh reviewed (commented):
re-did merge to honor refactor in https://github.com/ray-project/ray/pull/57181
addressed comments + re-running tests

kouroshHakha reviewed (commented):
cool. just one modification needed.

nrghosh committed:
move DP placement strategy logic to DPServer

- and update tests

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>
